### PR TITLE
Reorganize the parser

### LINKED ISSUE #236: Parser doesn't allow the binders in types to be wrapped in parens.
At the top level, creating triangular arrays works fine:
```
n = Fin 4
:t for j:n. for i:(..j). 1.0
((j:(Fin 4)) => (..j) => Float32)
```

But not inside a function or declaration:
```
func : _ =
  for j:n. for i:(..j). 1.0
Type error:Leaked type variable: j:?1

  for j:n. for i:(..j). 1.0
  ^^^^
```

It also doesn't appear to be possible to spell the type of a function that returns a triangular array, although it can be inferred if you use a ```_```.  Copying the output of ```:t``` above gives:
```
def func2 (a:Type) : ((c:a)=>(..c)=>Float32) =
  for j:a. for i:(..j). 1.0
Parse error:7:27:
  |
7 | def func2 (a:Type) : ((c:a)=>(..c)=>Float32) =
  |                           ^^^
unexpected ")=>"
expecting ''', '_', alphanumeric character, or arrow
```

While we're here, I also wish the "Leaked type variable" error message gave more details.

### LINKED ISSUE #953: RFC Parsing reorg
Let us view parsing a Dex program as a sequence of five stages:
1. Blocking
2. Lexing
3. Grouping (matching parens, resolving indentation-induced groups)
4. Precedence resolution
5. Abstract syntax.

Currently, the latter four stages are fused, in `Parser.hs`, into a single function `:: Parser SourceBlock'`.  Perhaps we could have simpler code and a more uniform surface syntax if we split the last stage off, by introducing an intermediate representation for the result of the first four stages.  The intermediate would represent the grouping structure of the code, without putting any other interpretation on it yet:
```
data Group = WithSrc Group'
data Group'
  = Ident String
  | Number Int -- Float, Nat, etc
  | Bracket Bracket Group
  | Brackets Bracket [Group]
  | Binary Group Group Group -- could encode whitespace-as-application here or add another constructor for it
  | Unary Group Group -- covers unary - and unary +
  | Def Group [Group] Group GBlock -- name, parameters, return, body
  | For [Group] GBlock -- also for_, rof, rof_
  | Case Group [(Group, GBlock)] -- scrutinee, alternatives
  | If Group GBlock GBlock
  | Lambda [Group] GBlock
  -- etc for Data, Interface, Instance, ...

data Bracket = Parens | Square | Curly -- and we can add others, like @{ or [| or whatever

data GBlock = GBlock [(Group, Group)] (Maybe Group) -- decls and optional tail expression
```

With this defined, we de-fuse `Parser SourceBlock'` into a `grouping :: Parser GBlock` and another function `abstractSyntax :: GBlock -> m SourceBlock'`.  The errors that are natural to detect in the former would be relatively basic things like forgetting to close a paren or a doublequote, but it would accept nonsense like `def frotz (x = a + b : Type) : Float = <body>`, and it would be up to `abstractSyntax` to complain about that.

What would we gain from this rewrite?
1. We could easily allow grouping parens everywhere, and fix #236.
2. We could make ast errors easier to debug by showing the grouping structure in hovertips.
3. This _might_ make grouping errors easier to debug too, by not conflating them with stuff that could affect the ast interpretation.
4. The parsing ambiguity between a table literal and an interface constraint (both in square brackets) would become irrelevant and we could easily fix #925.
5. `grouping` _should_ be faster than the current parser because it shouldn't need to backtrack as much, but that might be offset by the cost of the second traversal.  On the other hand, that second traversal should be relatively cheap, because it would operate on a more compressed representation.  Then again, this gain is probably best captured by just pulling out a separate lexer using a fast regex engine.

Thoughts?

### PR DESCRIPTION
**Pull request by axch**
This change

- Defines a new intermediate ADT for concrete syntax, which,
  as proposed in Issue #953, captures grouping structure but
  is much looser than the proper Dex ADT.

- Splits the parser into three new parsing files named Lexing.hs,
  ConcreteSyntax.hs, and AbstractSyntax.hs.

- Moves the current Megaparsec-based lexer to Lexing.hs more or less
  unchanged.

- Moves everything from Parser.hs up to the new concrete syntax ADT to
  ConcreteSyntax.hs, which imports Lexing.hs.

- Moves the rest of Parser.hs to AbstractSyntax.hs.  AbstractSyntax.hs
  imports ConcreteSyntax.hs, and exposes the same API that Parser.hs
  used to expose.

- Updates the corpus and documentation to the new syntax.

- Closes #953.

The performance impact on our benchmarks from this change appears to
be minimal.

While the goal for this change was to be as backward-compatible as
possible (deferring intentional syntax changes to separate PRs), the
surface syntax of Dex changes in at least the following ways:

- Dependent types such as the triangular array (i:Fin n)=>(..i) now
  require parens around the binder instead of forbidding them.
  Mis-parses of `:` produce an error message that tries to be helpful.
  - This fixes #236.

- Primitives allow operators (except juxtaposition and indexing)
  in their argument expressions.

- Non-parenthesized patterns are allowed in all argument-definition
  positions (including def).  This fixes https://github.com/google-research/dex-lang/issues/955.

- Lambda args cannot directly be record, table, or tuple patterns
  now; but you can get a pattern by nesting: `\((y, x)). body`
  - Is this worth changing?

- Instead, curly braces `{}`, square brackets `[]`, and parens `()` in
  lambda argument declarations mean the same thing they mean in `def`,
  namely, implicit, class, and explicit arguments, respectively.
  - This fixes https://github.com/google-research/dex-lang/issues/954.

- Indented blocks must now actually be indented (by a positive number
  of spaces).

- Records used to allow fields like `foo=\x y.stuff` with no space
  between the `=` and the `\`.  That space is now required.

- {|} (without spaces) used to parse as an empty variant type.
  Now |} is always a single lexeme, so the space is required.

- Infix operators are now allowed as class methods (whether they have
  pre-defined precedence or not).

- The parser now allows parens (meaning grouping) everywhere,
  except the places where they have a specific other meaning:
  - At the top level under a `def` or `lambda`, the first level of
    parens means "this is an explicit argument".  In particular,
    ({n}) means "match a one-field record with field name `n` and
    bind the value to variable `n`".
  - At the top level of an instance definition, the first level of
    parens means "this is an argument", as opposed to the class
    name for which this is an instance.
  - Around a single symbol, parens mean "treat this as a prefix
    function name instead of an infix operator".
  - Effect row parsing, like `{State h, (Read h | effs)}`.  The parens
    are forbidden from hiding the `|`.
  - In case patterns, parens mean "the first item is a constructor to match".
  - Note that grouping can change the meaning of other operators,
    especially juxtaposition
    - The argument `{foo bar baz}` means three implicit arguments,
      `foo`, `bar`, and `baz`; whereas `{(foo bar baz)}` means one
      implicit argument, which is a pattern for the `foo` constructor.
    - The type annotation `: ({IO}) a` prevents {IO} from being
      interpreted as an effect.
    - There are probably other examples.

- In record patterns, a bare ellipsis used to be syntactic sugar
  for ignoring the remaining fields.  Now, that has to be spelled
  with an explicit underscore, as `{foo, bar, ..._}`.

- Separators of curly-braced lists used to be treated specially, such
  that `{|...rest}` would lex as `{ | ... rest }`.  Now they are
  symbols, so `|...` is a single lexeme; to achieve the old effect,
  include a space: `{| ...rest}`.

- Likewise, in record fields, `x=-x` used to work, but now requires a
  space: `x= -x` or `x = -x`.

Remaining to do:

- https://github.com/google-research/dex-lang/issues/954,
- https://github.com/google-research/dex-lang/issues/925,
- https://github.com/google-research/dex-lang/issues/997,
- and anything else in the "language / parser" label.

### CODE REVIEW

File: doc/functions.dx [hunk 1]
```diff
@@ -242,23 +242,16 @@ them in parens like `def`.
 map (\(a:Float). a + 1) [1., 2., 3.]
 > [2., 3., 4.]
 
-'Like `def`, we can also specify implicit parameters with curly braces,
+'Also like `def`, we can specify implicit parameters with curly braces:
 
 \{a} (x:a). x
 > \a:Type ?-> \x:a. x
```
Comment (apaszke) [inline doc/functions.dx:248]: Should we adjust the pretty printer to include those parens? This type looks like it wouldn't parse with your changes.

File: doc/functions.dx [hunk 2]
```diff
@@ -242,23 +242,16 @@ them in parens like `def`.
...
->       ^^^^
-
-'The workaround is to either use a `def`, or to annotate the type of
-the anonymous function.  If the type include interface constraints,
-the compiler will insert an appropriate parameter to accept them.
+> \n:Type ?->
+>   \v#0:(Ix n) ?=>
+>     \xs:(n => Float32).
+>       v#1:Float32 = sum n Float32 (Instance Add.2 []) v#0 xs
+>       v#1
```
Comment (apaszke) [inline doc/functions.dx:257]: Perhaps print the type of this function instead of its implementation?

Comment (apaszke) [inline doc/functions.dx:257]: Perhaps print the type of this function instead of its implementation?

File: examples/quaternions.dx [hunk 3]
```diff
@@ -23,8 +23,8 @@ q
 ' ## Quaternion magnitude
  A quaternion is basically a 4 dimensional vector.
 
-as_vec = \Q{x,y,z,w}. [x, y, z, w]
-from_vec = \[x, y, z, w]. Q{x,y,z,w}
+as_vec = \(Q{x,y,z,w}). [x, y, z, w]
```
Comment (apaszke) [inline examples/quaternions.dx:26]: Do _all_ patterns have to be wrapped in parens?

File: python/dex/interop/jax/apply.py [hunk 4]
```diff
@@ -334,10 +334,16 @@ def tuple_arg_string(names, name_to_ty):
   return f"({tuple_ref_string(names)} : {ty})"
 
 def tuple_ref_string(names):
-  return "(" + ", ".join(names) + ")"
+  if len(names) == 1:
+    return names[0]
+  else:
+    return "(" + ", ".join(names) + ")"
```
Comment (apaszke) [inline python/dex/interop/jax/apply.py:340]: Why is this change necessary?

File: tests/record-variant-tests.dx [hunk 5]
```diff
@@ -84,39 +84,31 @@ Syntax for records, variants, and their types.
...
 
 :p {a:Int & b:Float | c:Int }
-
-> Parse error:94:21:
->    |
-> 94 | :p {a:Int & b:Float | c:Int }
->    |                     ^
-> unexpected '|'
-> expecting "..", "::", "<..", "=>", "{|", '$', '&', '.', '}', arrow, backquoted name, expression, infix operator, or resume
+> Syntax error: Bad field.  Expected `label : expr`,`... expr`, or a single `label`.
```
Comment (apaszke) [inline tests/record-variant-tests.dx:95]: I'm not sure if I understand this error message. Can we try to clarify it?

File: src/lib/TopLevel.hs [hunk 6]
```diff
@@ -335,15 +335,17 @@ evalSourceBlock' mname block = case sbContents block of
...
-  GetNameType v -> do
-    ty <- sourceNameType v
-    logTop $ TextOut $ pprintCanonicalized ty
-  ImportModule moduleName -> importModule moduleName
-  QueryEnv query -> void $ runEnvQuery query $> UnitE
-  ProseBlock _ -> return ()
-  CommentLine  -> return ()
-  EmptyLines   -> return ()
   UnParseable _ s -> throw ParseErr s
+  BadSyntax e -> throwErrs e
```
Comment (apaszke) [inline src/lib/TopLevel.hs:339]: What's the difference between UnParseable and BadSyntax?

Comment (apaszke) [inline src/lib/TopLevel.hs:339]: What's the difference between UnParseable and BadSyntax?

File: src/lib/Types/Source.hs [hunk 7]
```diff
@@ -307,13 +309,18 @@ data SourceBlock' =
...
- | QueryEnv EnvQuery
- | ProseBlock Text
- | CommentLine
- | EmptyLines
+ | Misc SourceBlockMisc
  | UnParseable ReachedEOF String
+ | BadSyntax Errs
+  deriving (Show, Generic)
+
+data SourceBlockMisc
```
Comment (apaszke) [inline src/lib/Types/Source.hs:317]: Not sure if I'm a fan of this misc category. How do we decide what goes in there?

**Review by apaszke — APPROVED** [review]
Could you please explain why we need the space between parens and contents (in records and variants)?

LGTM, but I have barely looked at the code. The diff is so large that I'm unable to go through it.

**Review by apaszke — APPROVED** [review]
Could you please explain why we need the space between parens and contents (in records and variants)?

LGTM, but I have barely looked at the code. The diff is so large that I'm unable to go through it.

File: doc/functions.dx (hunk 1 above)
Comment (axch) [reply doc/functions.dx:248]: Come to think of it, that was printing the function (not its type), and it looks like it wouldn't have parsed before, either.  Anyway, changed the pretty-print of lambda terms to emit the new syntax.

File: doc/functions.dx (hunk 2 above)
Comment (axch) [reply doc/functions.dx:257]: Done.

File: python/dex/interop/jax/apply.py (hunk 4 above)
Comment (axch) [reply python/dex/interop/jax/apply.py:340]: Because without it, the above would emit stuff like `\((thing):(type)). stuff`, trying to imply a one-element tuple.  It seemed easier to avoid the confusion with grouping parens at this level.

File: examples/quaternions.dx (hunk 3 above)
Comment (axch) [reply examples/quaternions.dx:26]: In lambda and def arguments, currently yes.

In this case, we would need to disambiguate between the intent, which is a pattern matching the `Q` constructor with `x, y, z, w` fields, and an alternate interpretation as an explicit argument named `Q` followed by a list of implicit arguments named `x, y, z, w`.  The parens serve to do that.  Though we could certainly consider a more elaborate disambiguation rule.

File: tests/record-variant-tests.dx (hunk 5 above)
Comment (axch) [reply tests/record-variant-tests.dx:95]: Done.

File: src/lib/TopLevel.hs (hunk 6 above)
Comment (axch) [reply src/lib/TopLevel.hs:339]: `BadSyntax` is a failure at the abstract syntax stage.  Comment added to the `SourceBlock'` ADT.

File: src/lib/Types/Source.hs (hunk 7 above)
Comment (axch) [reply src/lib/Types/Source.hs:317]: This is the stuff that requires no concrete->abstract syntax interpretation (and isn't a parse error).  Comment added.

**Comment by axch** [comment]
Re: space between parens and contents: because the lexical syntax is more uniform now.  From the main lexer's point of view, `|}` is and has always been one lexeme; however, `{` used to, in effect, introduce a sub-parser that specially detected `|`, such that `{|}` (no spaces) would parse the same as `{ | }` (a variant with no alternatives).  That sub-parser is gone now, and the stuff following a `{` is lexed and parsed the same way as everything else; so if you want `| }` to be two lexemes instead of one, you have to write the space there.

The same applies to the other field labeler and field separator characters: `x=\foo. bar` used to (perhaps unintentionally?) parse as a record field because of special treatment of `=` when parsing records, but now doesn't because `=\` registers as a lexeme (as it would outside of a record context).

**Comment by axch** [comment]
@dougalm : Merging now to avoid staleness, but would definitely welcome any retroactive code review you might wish to give.
